File: study/lua/src/app/scene/view/instrumentDetail/InstrumentDetailAttrNode.lua
--
-- Author: Liangxu
-- Date: 2017-9-16 14:13:03
--
local ListViewCellBase = require("app.ui.ListViewCellBase")
local InstrumentDetailAttrNode = class("InstrumentDetailAttrNode", ListViewCellBase)
local UserDataHelper = require("app.utils.UserDataHelper")
local TextHelper = require("app.utils.TextHelper")
local InstrumentTrainHelper = require("app.scene.view.instrumentTrain.InstrumentTrainHelper")
local AttributeConst = require("app.const.AttributeConst")
local InstrumentConst = require("app.const.InstrumentConst")
local LogicCheckHelper = require("app.utils.LogicCheckHelper")

function InstrumentDetailAttrNode:ctor(instrumentData, rangeType)
	self._instrumentData = instrumentData
	self._rangeType = rangeType

	local resource = {
		file = Path.getCSB("InstrumentDetailAttrNode", "instrument"),
		binding = {
			_buttonAdvance = {
				events = {{event = "touch", method = "_onButtonAdvanceClicked"}}
			},
			_buttonLimit = {
				events = {{event = "touch", method = "_onButtonLimitClicked"}}
			},
		},
	}
	InstrumentDetailAttrNode.super.ctor(self, resource)
end

function InstrumentDetailAttrNode:onCreate()
	local contentSize = self._panelBg:getContentSize()
	self:setContentSize(contentSize)
	self._nodeTitle:setFontSize(24)
	self._nodeTitle:setTitle(Lang.get("instrument_detail_title_attr"))
	self._buttonAdvance:setString(Lang.get("instrument_btn_advance"))
	self._buttonLimit:setString(Lang.get("instrument_btn_limit"))

	local des = Lang.get("instrument_detail_advance_level")
	local value = self._instrumentData:getLevel()
	local max = self._instrumentData:getAdvanceMaxLevel()
	local color = value < max and Colors.BRIGHT_BG_ONE or Colors.BRIGHT_BG_GREEN
	self._nodeLevel:setFontSize(20)
	self._nodeLevel:updateUI(des, value, max)
	self._nodeLevel:setValueColor(color)
	self._nodeLevel:setMaxColor(color)

	self:_updateAttrDes()

	local isUser = self._instrumentData:isUser()
	self._buttonAdvance:setVisible(isUser)

	local isCanLimit = self._instrumentData:isCanLimitBreak()
	local isShow = LogicCheckHelper.funcIsShow(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2)
	-- self._buttonLimit:setVisible(isUser and isCanLimit and isShow)
	self:_showLimitBtnProc()
	if isUser then
		local RedPointHelper = require("app.data.RedPointHelper")
		local reach1 = RedPointHelper.isModuleSubReach(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE1, "slotRP", self._instrumentData)
		self._buttonAdvance:showRedPoint(reach1)
		local reach2 = RedPointHelper.isModuleSubReach(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2, "slotRP", self._instrumentData)
		self._buttonLimit:showRedPoint(reach2)
	end
end

function InstrumentDetailAttrNode:_updateAttrDes()
	local attrInfo = UserDataHelper.getInstrumentAttrInfo(self._instrumentData)
	local desInfo = TextHelper.getAttrInfoBySort(attrInfo)
	for i = 1, 4 do
		local info = desInfo[i]
		if info then
			self["_nodeAttr"..i]:updateView(info.id, info.value, nil, 4)
			self["_nodeAttr"..i]:setVisible(true)
		else
			self["_nodeAttr"..i]:setVisible(false)
		end
	end
end

function InstrumentDetailAttrNode:_onButtonAdvanceClicked()
	local isOpen, des = LogicCheckHelper.funcIsOpened(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE1)
	if not isOpen then
		G_Prompt:showTip(des)
		return
	end

	local instrumentId = self._instrumentData:getId()
	G_SceneManager:showScene("instrumentTrain", instrumentId, InstrumentConst.INSTRUMENT_TRAIN_ADVANCE, self._rangeType, true)
end

function InstrumentDetailAttrNode:_onButtonLimitClicked()
	-- FUNC_INSTRUMENT_TRAIN_TYPE2_RED
	local isOpen, des = LogicCheckHelper.funcIsOpened(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2)
	if not isOpen then
		G_Prompt:showTip(des)
		return
	end

	local instrumentId = self._instrumentData:getId()
	G_SceneManager:showScene("instrumentTrain", instrumentId, InstrumentConst.INSTRUMENT_TRAIN_LIMIT, self._rangeType, true)
end

function InstrumentDetailAttrNode:_showLimitBtnProc()
	local isUser = self._instrumentData:isUser()
	local isCanLimit = self._instrumentData:isCanLimitBreak()
	local isShow = self._instrumentData:getLimitFuncShow()
	self._buttonLimit:setVisible(isUser and isCanLimit and isShow)
end

return InstrumentDetailAttrNode

File: study/ts/script/scene/view/instrument/InstrumentDetailAttrNode.ts
const { ccclass, property } = cc._decorator;

import ListViewCellBase from "../../../ui/ListViewCellBase";
import CommonButtonLevel2Highlight from "../../../ui/component/CommonButtonLevel2Highlight";
import { G_SceneManager, Colors, G_Prompt } from "../../../init";
import InstrumentConst from "../../../const/InstrumentConst";
import { handler } from "../../../utils/handler";
import { Lang } from "../../../lang/Lang";
import { InstrumentUnitData } from "../../../data/InstrumentUnitData";
import CommonDesValue from "../../../ui/component/CommonDesValue";
import { LogicCheckHelper } from "../../../utils/LogicCheckHelper";
import { RedPointHelper } from '../../../data/RedPointHelper';
import { FunctionConst } from "../../../const/FunctionConst";
import { TextHelper } from "../../../utils/TextHelper";
import { InstrumentDataHelper } from "../../../utils/data/InstrumentDataHelper";
import CommonAttrNode from "../../../ui/component/CommonAttrNode";
import CommonDetailTitleWithBg from "../../../ui/component/CommonDetailTitleWithBg";

@ccclass
export default class InstrumentDetailAttrNode extends ListViewCellBase {

    @property({
        type: CommonButtonLevel2Highlight,
        visible: true
    })
    _buttonAdvance:CommonButtonLevel2Highlight = null;

    @property({
        type: CommonButtonLevel2Highlight,
        visible: true
    })
    _buttonLimit:CommonButtonLevel2Highlight = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBg:cc.Node = null;

    @property({
        type: CommonDetailTitleWithBg,
        visible: true
    })
    _nodeTitle:CommonDetailTitleWithBg = null;

    @property({
        type: CommonDesValue,
        visible: true
    })
    _nodeLevel:CommonDesValue = null;

    @property({
        type: CommonAttrNode,
        visible: true
    })
    _nodeAttr1:CommonAttrNode = null;

    @property({
        type: CommonAttrNode,
        visible: true
    })
    _nodeAttr2:CommonAttrNode = null;

    @property({
        type: CommonAttrNode,
        visible: true
    })
    _nodeAttr3:CommonAttrNode = null;

    @property({
        type: CommonAttrNode,
        visible: true
    })
    _nodeAttr4:CommonAttrNode = null;

    private _instrumentData:InstrumentUnitData;
    private _rangeType:number;

    onCreate(){
        var contentSize = this._panelBg.getContentSize();
        this.node.setContentSize(contentSize);
    }
    onEnter(){
        this._buttonAdvance.addTouchEventListenerEx(handler(this,this._onButtonAdvanceClicked),true);
        this._buttonLimit.addTouchEventListenerEx(handler(this,this._onButtonLimitClicked),true);

        this._nodeTitle.setFontSize(24);
        this._nodeTitle.setTitle(Lang.get('instrument_detail_title_attr'));
        this._buttonAdvance.setString(Lang.get('instrument_btn_advance'));
        this._buttonLimit.setString(Lang.get('instrument_btn_limit'));
        var des = Lang.get('instrument_detail_advance_level');
        var value = this._instrumentData.getLevel();
        var max = this._instrumentData.getAdvanceMaxLevel();
        var color = value < max && Colors.BRIGHT_BG_ONE || Colors.BRIGHT_BG_GREEN;
        this._nodeLevel.setFontSize(20);
        this._nodeLevel.updateUI(des, value, max);
        this._nodeLevel.setValueColor(color);
        this._nodeLevel.setMaxColor(color);
        this._updateAttrDes();
        var isUser = this._instrumentData.isUser();
        this._buttonAdvance.setVisible(isUser);
        //var isCanLimit = this._instrumentData.isCanLimitBreak();
        // var isShow = LogicCheckHelper['funcIsShow'](FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2);
        // this._buttonLimit.setVisible(isUser && isCanLimit && isShow);
        this._showLimitBtnProc();
        if (isUser) {
            var reach1 = RedPointHelper.isModuleSubReach(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE1, 'slotRP', this._instrumentData);
            this._buttonAdvance.showRedPoint(reach1);
            var reach2 = RedPointHelper.isModuleSubReach(FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2, 'slotRP', this._instrumentData);
            this._buttonLimit.showRedPoint(reach2);
        }
    }

    init(instrumentData, rangeType){
        this._instrumentData = instrumentData;
        this._rangeType = rangeType;
    }

    _onButtonAdvanceClicked() {
        var result = LogicCheckHelper['funcIsOpened'](FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE1);
        var isOpen = result[0], des = result[1];
        if (!isOpen) {
            G_Prompt.showTip(des);
            return;
        }
        var instrumentId = this._instrumentData.getId();
        G_SceneManager.showScene('instrumentTrain', instrumentId, InstrumentConst.INSTRUMENT_TRAIN_ADVANCE, this._rangeType, true);
    }
    _onButtonLimitClicked() {
        var [isOpen, des] = LogicCheckHelper['funcIsOpened'](FunctionConst.FUNC_INSTRUMENT_TRAIN_TYPE2);
        if (!isOpen) {
            G_Prompt.showTip(des);
            return;
        }
        var instrumentId = this._instrumentData.getId();
       G_SceneManager.showScene('instrumentTrain', instrumentId, InstrumentConst.INSTRUMENT_TRAIN_LIMIT, this._rangeType, true);
    }
    _updateAttrDes(){
        var attrInfo = InstrumentDataHelper.getInstrumentAttrInfo(this._instrumentData);
        var desInfo = TextHelper.getAttrInfoBySort(attrInfo);
        for (var i = 1; i<=4; i++) {
            var info = desInfo[i-1];
            if (info) {
                this['_nodeAttr' + i].updateView(info.id, info.value, null, 4);
                this['_nodeAttr' + i].node.active = (true);
            } else {
                this['_nodeAttr' + i].node.active = (false);
            }
        }
    }

    _showLimitBtnProc() {
        var isUser = this._instrumentData.isUser();
        var isCanLimit = this._instrumentData.isCanLimitBreak();
        var isShow = this._instrumentData.getLimitFuncShow();
        this._buttonLimit.node.active = (isUser && isCanLimit && isShow);
    }
}
